fuzz-tests/ntt.cpp

```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
 }
 
 typedef vector<ll> vl;
-void ntt(ll* x, ll* temp, ll* roots, int N, int skip) {
-	if (N == 1) return;
-	int n2 = N/2;
-	ntt(x     , temp, roots, n2, skip*2);
-	ntt(x+skip, temp, roots, n2, skip*2);
-	rep(i,0,N) temp[i] = x[i*skip];
-	rep(i,0,n2) {
-		ll s = temp[2*i], t = temp[2*i+1] * roots[skip*i];
-		x[skip*i] = (s + t) % mod; x[skip*(i+n2)] = (s - t) % mod;
-	}
-}
-void ntt(vl& x, bool inv = false) {
-	ll e = modpow(root, (mod-1) / sz(x));
-	if (inv) e = modpow(e, mod-2);
-	vl roots(sz(x), 1), temp = roots;
-	rep(i,1,sz(x)) roots[i] = roots[i-1] * e % mod;
-	ntt(&x[0], &temp[0], &roots[0], sz(x), 1);
+int mul(int a, int b) { return (ll)a * b % mod; }
+int sub(int a, int b) { return b > a ? a - b + mod : a - b; }
+int add(int a, int b) { return a + b >= mod ? a + b - mod : a + b; }
+void ntt(vi &a, vi &rt, vi &rev, int n) {
+    rep(i, 0, n) if (i < rev[i]) swap(a[i], a[rev[i]]);
+    for (int k = 1; k < n; k *= 2)
+        for (int i = 0; i < n; i += 2 * k)
+            rep(j, 0, k) {
+                int z = mul(rt[j + k], a[i + j + k]);
+                a[i + j + k] = sub(a[i + j], z);
+                a[i + j] = add(a[i + j], z);
+            }
 }
-vl conv(vl a, vl b) {
-	int s = sz(a) + sz(b) - 1; if (s <= 0) return {};
-	int L = s > 1 ? 32 - __builtin_clz(s - 1) : 0, n = 1 << L;
-	if ((rand()%2 == 0) && s <= 200) { // (factor 10 optimization for |a|,|b| = 10)
-		vl c(s);
-		rep(i,0,sz(a)) rep(j,0,sz(b))
-			c[i + j] = (c[i + j] + a[i] * b[j]) % mod;
-		trav(x, c) if (x < 0) x += mod;
-		return c;
-	}
-	a.resize(n); ntt(a);
-	b.resize(n); ntt(b);
-	vl c(n); ll d = modpow(n, mod-2);
-	rep(i,0,n) c[i] = a[i] * b[i] % mod * d % mod;
-	ntt(c, true); c.resize(s);
-	trav(x, c) if (x < 0) x += mod;
-	return c;
+
+vl conv(const vi &a, const vi &b) {
+    if (a.empty() || b.empty())
+        return {};
+    vl res(sz(a) + sz(b) - 1);
+    int L = 32 - __builtin_clz(sz(res)), n = 1 << L;
+    vi inl(n), inr(n);
+    vi out(n), rt(n, 1);
+    vi rev(n);
+    rep(i, 0, n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+    int curL = (mod - 1) >> 2;
+    for (int k = 2; k < n; k *= 2) {
+        int z[] = {1, modpow(root, curL)};
+        curL >>= 1;
+        rep(i, k, 2 * k) rt[i] = mul(rt[i / 2], z[i & 1]);
+    }
+    copy(all(a), begin(inl)), copy(all(b), begin(inr));
```

Comment:
an alternative to this would be
```
vi inl(all(a)), inr(all(b));
inl.resize(n); inr.resize(n);
```
(it adds some memory allocations but hopefully it's not that much slower). OTOH that might not be a LOC win when putting all `vi` declarations on the same line.

Reply:
You can also just do `vi inl(a), inr(b)`, right? so it's also a character win, at least.